### PR TITLE
code refactoring to provide abstracted library

### PR DESCRIPTION
**Pull request by xm1k3**
```go
func main() {
	// Parse the command line flags and read config files
	options := runner.ParseOptions()

	agentFactory, err := runner.New(options)
	if err != nil {
		gologger.Fatal().Msgf("Could not create runner: %s\n", err)
	}
	newRunner, err := runner.NewRunner(options, agentFactory)
	if err != nil {
		gologger.Fatal().Msgf("Could not create runner: %s\n", err)
	}

	err = newRunner.Run(context.Background())
	if err != nil {
		gologger.Fatal().Msgf("Could not run enumeration: %s\n", err)
	}
}
```

### CODE REVIEW

**Comment by tarunKoyalwar** [comment]
```console
$ go run . -s "ssl:'hackerone.com'" -v

  __  ______  _________ _   _____  _____
 / / / / __ \/ ___/ __ \ | / / _ \/ ___/
/ /_/ / / / / /__/ /_/ / |/ /  __/ /    
\__,_/_/ /_/\___/\____/|___/\___/_/

		projectdiscovery.io

[INF] Current uncover version v1.0.3 (latest)
[shodan] 2606:4700::6810:6434:2083
[shodan] 104.16.100.52:8443
[shodan] 104.16.99.52:8443
[shodan] 2606:4700::6810:6434:2087
[shodan] 2606:4700::6810:6334:443
[shodan] 2606:4700::6810:6334:8443
[shodan] 2606:4700::6810:6334:2087
[shodan] 104.16.99.52:2083
[shodan] 104.16.99.52:443
[shodan] 104.16.100.52:443
[shodan] 104.16.100.52:2083
[shodan] 2606:4700::6810:6434:8443
[shodan] 2606:4700::6810:6434:443
[shodan] 2606:4700::6810:6334:2083
[shodan] 164.92.231.194:443
```

**Comment by tarunKoyalwar** [comment]
### Integration / Library Code

- Now uncover is completely abstracted i.e logic/code does not need to be changed downstream (nuclei etc) if we add new/sources 

https://github.com/projectdiscovery/uncover/blob/99e091ac33f65bc490ca33a87c226d977561f7ca/examples/main.go#L1-L36

**Comment by tarunKoyalwar** [comment]
### Follow up Issue

- replace https://github.com/projectdiscovery/nuclei/blob/main/v2/pkg/protocols/common/uncover/uncover.go with ^

File: runner/options.go [hunk 1]
```diff
@@ -300,11 +227,57 @@ func (options *Options) validateOptions() error {
...
-	for _, engine := range options.Engine {
-		if strings.EqualFold(engine, "shodan-idb") {
-			return true
-		}
+func versionCallback() {
+	gologger.Info().Msgf("Current Version: %s\n", version)
+	os.Exit(0)
+}
+
+func appendAllQueries(options *Options) {
```
Comment (Mzack9999) [inline runner/options.go:235]: What do you think about the following:
```go
func appendQuery(options *Options, name string, queries ...string) {
	if len(queries) > 0 {
		options.Engine = append(options.Engine, name)
		options.Query = append(options.Query, queries...)
	}
}

func appendAllQueries(options *Options) {
	appendQuery(options, "shodan", options.Shodan...)
...
}
```

File: runner/runner.go [hunk 2]
```diff
@@ -34,197 +20,72 @@ func init() {
...
-						}
-					}
+	return runner, nil
+}
 
-				}
-			}(agent, uncoverQuery)
+// RunEnumeration runs the subdomain enumeration flow on the targets specified
+func (r *Runner) Run(ctx context.Context) error {
+
```
Comment (Mzack9999) [inline runner/runner.go:56]: remove new line

**Review by Mzack9999 — CHANGES_REQUESTED** [review]
Implementation: lgtm
Notes:
- suggesting minor optional refactor
- I think we should make it clear in the readme that now the runner is not reusable anymore with different queries, but a new instance needs to be created with new options

**Review by Mzack9999 — APPROVED** [review]
lgtm

**Review by ehsandeep — CHANGES_REQUESTED** [review]
This works

```
echo jira | ./uncover -v
```

but not this:

```
echo jira | ./uncover
```

**Comment by tarunKoyalwar** [comment]
```console
$ ./uncover -q "ssl:hackerone.com" -e shodan  -v

  __  ______  _________ _   _____  _____
 / / / / __ \/ ___/ __ \ | / / _ \/ ___/
/ /_/ / / / / /__/ /_/ / |/ /  __/ /    
\__,_/_/ /_/\___/\____/|___/\___/_/

		projectdiscovery.io

[INF] Current uncover version v1.0.4-dev (dev)
[shodan] 104.16.100.52:443
[shodan] 104.16.99.52:2083
[shodan] 104.16.100.52:2083
[shodan] 2606:4700::6810:6434:8443
[shodan] 2606:4700::6810:6434:2083
[shodan] 104.16.99.52:8443
[shodan] 104.16.100.52:8443
[shodan] 2606:4700::6810:6434:443
[shodan] 2606:4700::6810:6334:2083
[shodan] 2606:4700::6810:6334:443
[shodan] 104.16.99.52:443
[shodan] 2606:4700::6810:6334:8443
[shodan] 164.92.231.194:443
```
```console
$ ./uncover -q "ssl:hackerone.com" -e shodan 

  __  ______  _________ _   _____  _____
 / / / / __ \/ ___/ __ \ | / / _ \/ ___/
/ /_/ / / / / /__/ /_/ / |/ /  __/ /    
\__,_/_/ /_/\___/\____/|___/\___/_/

		projectdiscovery.io

[INF] Current uncover version v1.0.4-dev (dev)
104.16.100.52:443
104.16.99.52:2083
104.16.100.52:2083
2606:4700::6810:6434:8443
2606:4700::6810:6434:2083
104.16.99.52:8443
104.16.100.52:8443
2606:4700::6810:6434:443
2606:4700::6810:6334:2083
2606:4700::6810:6334:443
104.16.99.52:443
2606:4700::6810:6334:8443
164.92.231.194:443
```

**Review by ehsandeep — APPROVED** [review]

**Pull request merged**